addons/mail/models/res_users.py

```diff
@@ -273,19 +275,28 @@ def _init_messaging(self):
 
     @api.model
     def systray_get_activities(self):
-        activities = self.env["mail.activity"].search([("user_id", "=", self.env.uid)])
+        MailActivity = self.env["mail.activity"]
+        activities = MailActivity.search([("user_id", "=", self.env.uid)])
         activities_by_record_by_model_name = defaultdict(lambda: defaultdict(lambda: self.env["mail.activity"]))
         for activity in activities:
             record = self.env[activity.res_model].browse(activity.res_id)
             activities_by_record_by_model_name[activity.res_model][record] += activity
-        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_record_by_model_name.keys()})
-        user_activities = {}
+        activity_ids_without_access = self.env["mail.activity"]
+        activities_by_model_name = defaultdict(lambda: self.env["mail.activity"])
         for model_name, activities_by_record in activities_by_record_by_model_name.items():
-            domain = [("id", "in", list({r.id for r in activities_by_record.keys()}))]
+            res_ids = list({r.id for r in activities_by_record.keys()})
+            allowed_records = self._systray_get_allowed_records(model_name, res_ids)
+            for record, activities in activities_by_record.items():
+                if record not in allowed_records:
+                    activity_ids_without_access += activities
+                else:
+                    activities_by_model_name[model_name] += activities
+        if activity_ids_without_access:
+            activities_by_model_name['mail.activity'] = activity_ids_without_access
+        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_model_name.keys()})
```

Comment:
Keys won't have duplicates, we can use '_get_id', so seems we can simplify
```suggestion
        model_ids = [self.env["ir.model"]._get_id(name) for name in activities_by_model_name.keys()]
```